raft.go

```diff
@@ -927,14 +927,27 @@ func (r *raft) hasUnappliedConfChanges() bool {
 	if r.raftLog.applied >= r.raftLog.committed { // in fact applied == committed
 		return false
 	}
-	ents, err := r.raftLog.slice(r.raftLog.applied+1, r.raftLog.committed+1, noLimit)
-	if err != nil {
-		r.logger.Panicf("unexpected error getting unapplied entries (%v)", err)
-	}
-	for i := range ents {
-		if ents[i].Type == pb.EntryConfChange || ents[i].Type == pb.EntryConfChangeV2 {
-			return true
+	// Scan all unapplied committed entries to find a config change. Paginate the
+	// scan, to avoid a potentially unlimited memory spike.
+	//
+	// TODO(pavelkalinnikov): this scan does not need to happen each time. It's
```

Comment:
OTOH, this scan happens rarely (only before election), so overall it might be cheaper (and simpler) to do a rare enough scan than maintain this predicate through every state change. Maybe I should replace this TODO with a note why the scan is ok.

Reply:
There is a `raft.pendingConfIndex` field which might be already tracking what we need here.

Reply:
Agreed to discuss it separately.